Parallel interpolation

johnomotani opened the pull request:
Provides methods `BoutDataArray.getHighRes()` to get a version of a variable interpolated in the parallel direction to increase the poloidal resolution, or `BoutDataSet.getHighResVars(['var1', 'var2', ...)` to get a `Dataset` with high-resolution versions of a list of variables. An example from TORPEX simulations - before interpolation
![base](https://user-images.githubusercontent.com/3958036/77232893-9a3ccc00-6b9b-11ea-8512-21dd253b23dd.gif)
and after
![highres](https://user-images.githubusercontent.com/3958036/77232899-a032ad00-6b9b-11ea-98e2-59a9bd23693b.gif)

Also adds a feature to the tests - can use `@pytest.mark.long` to mark a test as long, in which case it is skipped by default. Long tests are run if `--long` argument is passed to pytest, and the Travis tests do run the long tests.

Includes #107, this PR will be only 1,384 additions and 363 deletions after that is merged.

pep8speaks commented:
Hello @johnomotani! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`conftest.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/conftest.py):

> [Line 4:1](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/conftest.py#L4): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 8:1](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/conftest.py#L8): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

* In the file [`xbout/boutdataarray.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataarray.py):

> [Line 390:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataarray.py#L390): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 442:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataarray.py#L442): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 446:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataarray.py#L446): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 476:5](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataarray.py#L476): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 508:9](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataarray.py#L508): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 514:21](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataarray.py#L514): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

* In the file [`xbout/boutdataset.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataset.py):

> [Line 141:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataset.py#L141): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 149:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/boutdataset.py#L149): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)

* In the file [`xbout/geometries.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py):

> [Line 75:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 86:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py#L86): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 118:28](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py#L118): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 174:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py#L174): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 185:17](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py#L185): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 188:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py#L188): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 89 characters)
> [Line 190:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py#L190): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 197:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/geometries.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)









* In the file [`xbout/region.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py):

> [Line 95:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L95): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 98:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L98): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 117:47](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L117): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 408:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L408): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 420:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L420): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 443:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L443): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 451:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L451): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 473:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L473): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 477:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L477): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 481:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L481): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 485:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L485): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 507:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L507): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 89 characters)
> [Line 558:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L558): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 562:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L562): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 566:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L566): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)
> [Line 570:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/region.py#L570): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 89 characters)

* In the file [`xbout/tests/test_boutdataarray.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py):

> [Line 57:73](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L57): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 60:88](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L60): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 63:88](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L63): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 66:88](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L66): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 69:97](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L69): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 72:97](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L72): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 75:97](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L75): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 78:97](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L78): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 109:74](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L109): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 112:89](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L112): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 115:89](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L115): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 118:89](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L118): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 121:98](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L121): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 124:98](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L124): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 127:98](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L127): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 130:98](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L130): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 135:77](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L135): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 174:77](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L174): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 210:77](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L210): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 245:77](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L245): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 288:79](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L288): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '('
> [Line 296:77](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L296): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 334:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L334): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 89 characters)
> [Line 336:77](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L336): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 352:77](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataarray.py#L352): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'

* In the file [`xbout/tests/test_boutdataset.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataset.py):

> [Line 11:9](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataset.py#L11): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 106:36](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataset.py#L106): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 181:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataset.py#L181): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 89 characters)
> [Line 187:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataset.py#L187): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 89 characters)
> [Line 320:90](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_boutdataset.py#L320): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 89 characters)

* In the file [`xbout/tests/test_geometries.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_geometries.py):

> [Line 26:46](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_geometries.py#L26): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','





* In the file [`xbout/tests/test_region.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_region.py):

> [Line 23:1](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_region.py#L23): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1296:1](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/tests/test_region.py#L1296): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file



* In the file [`xbout/utils.py`](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/utils.py):

> [Line 46:1](https://github.com/boutproject/xBOUT/blob/236402c480ac1535425dd9b209157761a2aaa814/xbout/utils.py#L46): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

TomNicholas commented on xbout/region.py:
```diff
@@ -0,0 +1,589 @@
+        if self.connection_upper is not None:
+            yu += myg
+
+        return slice(xi, xo), slice(yl, yu)
```
Why not return a dictionary of slices? If it's of the form `{'x': slice(xi, xo), 'y': slice(yl, yu)}` then it can be passed directly into `.isel()` elsewhere

TomNicholas commented on xbout/region.py:
```diff
@@ -0,0 +1,589 @@
+        return lower, upper
+
+
+def _get_topology(ds):
```
Is this code ever going to be reached for a `slab` topology? Or `s-alpha`?

johnomotani replied:
`s-alpha` will reach this and will be labelled as `limiter`, which is a bit more generic. We don't have a specific `slab` geometry at the moment, but if we had one it could actually call the `_create_regions_toroidal()` function, and would be set up as `core`, `sol` or `limiter` (depending on `ixseps1`). The parallel interpolation isn't really needed for a slab though, because the slab is orthogonal so the perpendicular plane is actually an x-z plane.

I think at the moment, we could just call `_create_regions_toroidal` for any BOUT++ simulation because it covers every topology currently possible, but it's conceivable that at some point in the far future there might be a more general topology that couldn't be specified by `ixseps*` and `jyseps*`...

TomNicholas replied:
`limiter` makes more sense. 

> if we had one it could actually call the _create_regions_toroidal() function, and would be set up as core, sol or limiter (depending on ixseps1).

This would be useful if anyone goes forward and implements e.g. region-aware differencing.

> The parallel interpolation isn't really needed for a slab though, because the slab is orthogonal so the perpendicular plane is actually an x-z plane.

I more just wanted to make sure it behaves gracefully if someone tries to use these methods for a slab geometry.

TomNicholas commented on xbout/region.py:
```diff
@@ -0,0 +1,589 @@
+        if region.connection_inner is not None:
+            if regions[region.connection_inner].connection_outer != region.name:
+                raise ValueError(
+                        'Inner connection of ' + region.name + ' is '
```
[f-strings ](https://realpython.com/python-f-strings/) would be neater

TomNicholas commented:
(Just realised I probably should have put these review comments on #107 instead...)

johnomotani commented on xbout/boutdataarray.py:
```diff
@@ -133,6 +133,401 @@ def fromFieldAligned(self):
+        result = xr.combine_by_coords(parts)
+        result.attrs = parts[0].attrs
+        # xr.combine_by_coords does not keep attrs at the moment. See
+        # https://github.com/pydata/xarray/issues/3865
```
https://github.com/pydata/xarray/pull/3877 :smiley:

TomNicholas replied:
:O

johnomotani commented on xbout/region.py:
```diff
@@ -0,0 +1,589 @@
+from collections import OrderedDict
```
We didn't need `OrderedDict`. I think I put them in because at one point I was considering accessing them by integer index sometimes, but didn't in the end. Will remove in this PR rather than #107 to avoid merge conflicts.

johnomotani commented on xbout/boutdataset.py:
```diff
@@ -78,6 +79,81 @@ def getFieldAligned(self, name, caching=True):
+            Factor to increase parallel resolution by
+        """
+
+        ds = self.data
```
I hadn't thought about it... Do we have any mechanism for global options/settings? For this one, I'd be inclined to keep as a `Dataset`-specific thing, on the basis that you need a certain factor for a certain configuration (e.g. machine, q95), so while you might want to apply the same one to several variables from the same `Dataset` (which is the purpose of this setting), you wouldn't generally want to apply the same one to all `Dataset`s. I don't have strong feelings about it though.

TomNicholas replied:
> Do we have any mechanism for global options/settings

We have (had?) a global dictionary of registered geometries, and options could similarly be done through a global dictionary (see for example in xarray).

> For this one, I'd be inclined to keep as a Dataset-specific thing, on the basis that you need a certain factor for a certain configuration (e.g. machine, q95),

Fair enough, probably makes sense to keep it tied to the object here then.

johnomotani commented on xbout/boutdataset.py:
```diff
@@ -78,6 +79,81 @@ def getFieldAligned(self, name, caching=True):
+        # Apply geometry
+        try:
+            ds = apply_geometry(ds, ds.geometry)
+        except AttributeError as e:
```
That is much nicer than checking `str(e)` to catch the particular exception!

johnomotani commented on xbout/boutdataset.py:
```diff
@@ -78,6 +79,81 @@ def getFieldAligned(self, name, caching=True):
+
+        return ds
+
+    def getHighParallelResVars(self, variables, **kwargs):
```
I'll rename this as `interpolate_parallel`. I am inclined to leave `variables` as a required argument though, because I don't want to encourage using something like
```
ds_highres = ds.interpolate_parallel()
```
to get *all* of the variables in high resolution. That would also need to load all of the high resolution variables into memory! I think it's better to guide the user to pass only the specific subset of variables that they really want in high resolution.

TomNicholas replied:
Fair enough. There should be a `variables='all'` special case though so I don't have to type out the whole list.

johnomotani replied:
Can we use something that's not a `str` for the special argument? It's unlikely that someone has a variable called `'all'` but not impossible. Would say `variables=True` be an abuse of `True`? It's not perfect, but I like that better than `variables=None` or `variables=False` for 'include everything'.

TomNicholas replied:
How about using the `Ellipsis` built-in? Then the syntax for all variables would be `variables=...`. See for example in [this xarray PR].

johnomotani replied:
Yes, that reads right, I like it! :+1:

johnomotani commented:
Note for myself: `BoutDataArray.interpolate_parallel()` does not create the high-resolution regions (at the moment they are only handled in `BoutDataset.interpolate_parallel()`), so the result does not support poloidal plots. Probably need to add a call to `apply_geometry()` in the method somewhere, with an optional argument to turn it off so `BoutDataset.interpolate_parallel()` doesn't create the regions multiple times.

TomNicholas commented on xbout/geometries.py:
```diff
@@ -118,84 +177,73 @@ def add_toroidal_geometry_coords(ds, *, coordinates=None, grid=None):
+        for v in needed_variables:
+            if v not in ds:
+                if grid is None:
+                    raise ValueError("Grid file is required to provide %s. Pass the "
```
Grid file is required to provide what?

johnomotani commented:
I've fixed the test failures now...

TomNicholas commented:
Is this now complete? If so, then tomorrow I might double check it doesn't break loading data for slab geometries then we could merge it?

johnomotani commented:
Yes, I think it's complete. I didn't think about slab geometry when writing though, I hope it's OK, but:
* `regions` are only created during `apply_geometry`
* in various places I've assumed that `regions` exist, in order to be 'generic'
    * hopefully only in functions that only make sense to use with a geometry, like plots with `plot_poloidal=True`, but I'll check
    * one example is that if for some reason you wanted to use `interpolate_parallel` in slab geometry, it would fail
* if it turns out to be easier to provide regions for slab geometry, that might be a reason for including a `geometry="slab"` option for `open_boutdataset`
* it would probably be a good idea to add a check to methods/functions that use `regions` to provide a helpful error message like "give a 'geometry' option to open_boutdataset" if the regions are missing.

todo (for myself):
- [ ] check if useful methods will fail for slab geometry because of missing `regions`
- [ ] add checks for missing `regions` with helpful error message (maybe by making `regions` a `@property`)

johnomotani commented:
`plot2d_wrapper`, which makes plots in the poloidal plane, will not work for slab geometry, but I think that's OK. `plot_separatrices` and `plot_targets` will also not work (not sure if they did before), although `plot_separatrix` is still OK without regions (because you have to pass it a value for `ixseps` yourself).

johnomotani commented:
Merge conflicts fixed now. Also updated some handling of `attrs`, because xarray-0.16.0 fixed some `attrs`-propagation allowing removal of some workarounds. In the process, I've made all 'coordinate' variables always have `"metadata"`, `"options"`, `"regions"` and `"geometry"` attrs, like the 'data_vars' variables do, which made things more consistent between a `BoutDataset` opened from `*.dmp.*.nc` files and one saved and re-loaded by xBOUT, simplifying the save-and-reload tests.

johnomotani commented:
That was a very strange issue. Tests failed on d062fa9 because they timed out. Turns out adding attrs to the `theta` coordinate made `.interp()` really slow in `BoutDataArray.interpolate_parallel()`. Adding a `da = da.compute()` before the `da.interp()` call fixed the slow-down. I guess adding attrs to the `theta` coordinate somehow messed up the dask task-graph, but don't even begin to understand how, or why calling `da = da.compute()` would fix it!

johnomotani commented:
As @TomNicholas noticed when discussing the performance regression issue today, using `da = da.compute()` forces computing and loading into memory the whole result, which we'd like to avoid if possible - especially if the variable being interpolated is too big to fit into memory after interpolation. I poked a bit more, and found that deleting the `"regions"` attr (which is a dict of `Region` objects, so I guess relatively big) on the 'theta' coordinate also fixed the performance regression. Coordinates always have their data in numpy arrays, not dask arrays. I still can't manage to make a simple reproducer, but apparently attrs on a numpy-backed variable passed to `dask.array.map_blocks()` (which is used down in the `xarray` `interp` methods) can slow things down by a factor of 10 (this was for a unit test case with stupidly small arrays and stupidly small chunks, so might not translate to a normal dataset, but there aren't that many chunks, and the test was taking ~35s in all, which seems high for the tiny array it's working on, although there are a lot of variables that I added as coordinates that also need interpolating). With the fix, the unit test only takes ~11s. Some of the cases of the same test (see below) with `vars_to_interpolate=...` (so all variables in the Dataset are interpolated) were taking ~6mins before the fix, and ~30s now.

@TomNicholas - I'd like to put this as an FYI issue on `xarray`, but without being able to reproduce for the devs there it seems like a waste of time. The case I've been checking is `time pytest --long -s -k test_interpolate_parallel[vars_to_interpolate0-False-False-guards0`, and commenting out this bit
https://github.com/boutproject/xBOUT/blob/66b286a84da16cf2d7bedd39553b4496a7071a44/xbout/boutdataarray.py#L314-L320
should re-introduce the slow-down if anyone has time to take a look.

TomNicholas commented:
Thanks for this report @johnomotani . This behaviour is definitely something I would like to understand, and ideally flag up with an issue on xarray. But for that I think it would need a reproducible example though. With this I can come back to it later at least.

Do you want me to review this so you can merge it and move on?

johnomotani commented:
> Do you want me to review this so you can merge it and move on?

That would be great! :+1:

johnomotani commented:
Tests pass now! Any more review comments before I merge?
